app/src/main/java/org/dicekeys/app/fragments/dicekey/SoloKeyFragment.kt

```diff
@@ -1,21 +1,230 @@
+
+    /**
+     * Check USB Device is ready to write or nor
+     */
+    fun deviceReadyToWrite(): Boolean =
+            deviceList.devices.values.firstOrNull()?.let {
+                deviceList.hasPermission(it)
+            } ?: false
+
+    /**
+     * Get Seed HEXAString
+     */
+    fun getSeed(): ByteArray? {
+        return try {
+            derivationRecipe.value?.let { Secret.deriveFromSeed(viewModel.diceKey.seed, it.derivationOptionsJson).secretBytes }
+        } catch (e: Throwable) {
+            null
+        }
+    }
+
+    /**
+     * Render attached usb devices and show ready to write view if device ready show view to write key
+     */
+    fun render() {
+        if (isSeedValid() &&
+                deviceReadyToWrite() &&
+                !isWriteUnderway) {
+            viewModel.soloKeyConnected()
+            deviceList.devices.values.firstOrNull()?.let {
+                binding.btnSidoKey.text = String.format(getString(R.string.seed_s_sn_s), it.productName, it.serialNumber)
+            }
+        }else{
+            //No solo key connected
+            viewModel.noSidoConnected()
+        }
+        timer.cancel()
+    }
+
+    /**
+     * Check attached usb and write seed to solo key
+     */
+    fun writeToCurrentSoloKey() {
+        if (deviceList.devices.size == 1) deviceList.devices.values.firstOrNull()?.let {
+            writeToSoloKey(it)
+        }
+    }
+
+    /**
+     * Write seed to solokey
+     */
+    private fun writeToSoloKey(device: UsbDevice) {
+        if (isWriteUnderway)
+            return
+        val seed = getSeed()
+        if (seed == null || !isSeedValid(seed))
+            return
+        isWriteUnderway = true
+
+        //Show Writing underway view
+        viewModel.writingUnderWay()
+        binding.tvSuccess.isVisible = false
+        binding.tvError.isVisible = true
+        timer.start()
+
+        Thread(Runnable {
+            try {
+                synchronized(this) {
+                    val connection = deviceList.connect(device)
+                    connection.loadKeySeed(seed)
+                }
+                isWriteUnderway = false
+
+                //Hide writing underway process and show connected Solo Key details with success
+                viewModel.soloKeyConnected()
+                requireActivity().runOnUiThread {
+                    isWriteUnderway = false
+                    binding.tvSuccess.isVisible = true
+                    render()
+                }
+
+
+            } catch (e: java.lang.Exception) {
+
+                //Solo Key wring failed
+                requireActivity().runOnUiThread {
+                    isWriteUnderway = false
+                    binding.tvError.isVisible = true
+                    binding.tvError.text = e.message
+                    render()
+                }
+            }
+        }).start()
+    }
+
+    /**
+     * Intialize countdown timer
+     */
+    val timer = object : CountDownTimer(2000000, 1000) {
```

Comment:
What are these hard-coded values?

(The current time limit from SoloKeys is 8 seconds.  That should be in a constant/parameter.

Reply:
First parameter is The number of millis in the future from the call to start() until the countdown is done
while second one is countdown interval in millis 

I'll change first parameter with content  values with 8 seconds

Reply:
> What are these hard-coded values?
> 
> (The current time limit from SoloKeys is 8 seconds. That should be in a constant/parameter.
First parameter is The number of millis in the future from the call to start() until the countdown is done
while second one is countdown interval in millis

I'll change first parameter with content values with 8 seconds